### PR TITLE
Extended Prepared Statement Support

### PR DESCRIPTION
**Pull request by anarazel**
This PR (which is based on #1146, and includes those changes) extends the prepared statement support to nearly all prepared statements, unless they're executed via SQL functions (in which case the amount of support should be unaffected).

To quote the commit message of how this is achieved:
```
    Hack up PREPARE/EXECUTE for nearly all distributed queries.
    
    All router, real-time, task-tracker plannable queries should now have
    full prepared statement support (and even use router when possible),
    unless they don't go through the custom plan interface (which
    basically just affects LANGUAGE SQL (not plpgsql) functions).
    
    This is achieved by forcing postgres' planner to always choose a
    custom plan, by assigning very low costs to plans with bound
    parameters (i.e. ones were the postgres planner replanned the query
    upon EXECUTE with all parameter values provided), instead of the
    generic one.
    
    This requires some trickery, because for custom plans to work the
    costs for a non-custom plan have to be known, which means we can't
    error out when planning the generic plan.  Instead we have to return a
    "faux" plan, that'd trigger an error message if executed.  But due to
    the custom plan logic that plan will likely (unless called by an SQL
    function, or because we can't support that query for some reason) not
    be executed; instead the custom plan will be chosen.
``` 

There's two major pre-requisite pieces in the first few commits of this PR:
1.  Support for deferred error messages.
```
    
    
    It can be useful, e.g. in the upcoming prepared statement support, to
    be able to return an error from a function that is not raised
    immediately, but can later be thrown.  That allows e.g. to attempt to
    plan a statment using different methods and to create good error
    messages in each planner, but to only error out after all planners
    have bene run.
    
    To enable that create support for deferred error messages that can be
    created (supporting errorcode, message, detail, hint) in one function,
    and then thrown in different place.
```
2.  Don't throw, but return, errors during router planning.
```
    Callers of router planner functionality now have to throw errors
    themselves if desired, but also can skip doing so.
    
    This is a pre-requisite for expanding prepared statement support.
    
    While touching all those lines, improve a number of error messages by
    getting them closer to the postgres error message guidelines.
```

There's some more cleanup needed, and some additional tests wouldn't hurt (but existing tests have been enabled/expanded). But functionality-wise I think this is how things should look like.

Fixes: #904
Fixes: #306

### CODE REVIEW

**Comment by anarazel** [comment]
@sumedhpathak @ozgune This is the, somewhat cleaned up, POC. I think we can make the explain stuff mergeable quite easily...  I think it's also quite straight-forward to add the hack that works for anything real-time plannable (even if it'd fail at execution time due to missing parameters!) due to point 3) above.   The larger support from 4) is a bit harder - I suspect within a day or two I could figure out a way to make this actually safe.

File: src/include/distributed/multi_router_planner.h [hunk 1]
```diff
@@ -28,10 +29,17 @@
 
 extern bool EnableRouterExecution;
 
-extern MultiPlan * MultiRouterPlanCreate(Query *originalQuery, Query *query,
-										 RelationRestrictionContext *restrictionContext);
+extern DeferredErrorMessage * CreateRouterPlan(MultiPlan **newPlan,
+											   Query *originalQuery, Query *query,
+											   RelationRestrictionContext *
+											   restrictionContext);
+extern DeferredErrorMessage * CreateModifyPlan(MultiPlan **newPlan,
```
Comment (marcocitus) [inline src/include/distributed/multi_router_planner.h]: Conceptually, returning a DeferredErrorMessage from high-level planning functions seems pretty strange. It would be nicer if these return a PlanningResult, which might contain a DeferredErrorMessage.

**Comment by anarazel** [comment]
Hi,

On 2017-01-19 21:16:10 -0800, Marco Slot wrote:
>  extern bool EnableRouterExecution;
>  
> -extern MultiPlan * MultiRouterPlanCreate(Query *originalQuery, Query *query,
> -										 RelationRestrictionContext *restrictionContext);
> +extern DeferredErrorMessage * CreateRouterPlan(MultiPlan **newPlan,
> +											   Query *originalQuery, Query *query,
> +											   RelationRestrictionContext *
> +											   restrictionContext);
> +extern DeferredErrorMessage * CreateModifyPlan(MultiPlan **newPlan,
> 
> Conceptually, returning a DeferredErrorMessage from high-level planning functions seems pretty strange. It would be nicer if these return a PlanningResult, which might contain a DeferredErrorMessage.

Hm - returning an error description or success from functions is a
pretty common pattern, isn't it?  It's e.g. all over linux and such?

**Comment by marcocitus** [comment]
> Hm - returning an error description or success from functions is a
> pretty common pattern, isn't it?  It's e.g. all over linux and such?

Most linux functions return success/failure and use errno. We could also use that pattern (i.e. use a global variable for the specific error that occurred), but PlanningResult seems better.

File: src/backend/distributed/planner/multi_planner.c [hunk 2]
```diff
@@ -117,38 +187,154 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
+	{
+		physicalPlan = CitusMakeNode(MultiPlan);
 
-		/* Create the physical plan */
-		physicalPlan = MultiPhysicalPlanCreate(logicalPlan);
+		if (!error)
+		{
+			error = DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								  "could not create distributed plan",
+								  NULL, NULL);
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_planner.c]: We can probably elaborate on the reason here.

**Comment by anarazel** [comment]
> Most linux functions return success/failure and use errno.

I mean the linux internal stuff, not posix/unix - there usually a negative value is returned to indicate the failure code (say ENOMEM), and a >= value to indicate success.

> We could also use that pattern (i.e. use a global variable for the specific error that occurred), but PlanningResult seems better.

I guess we can do that, but returning errors up the chain a consistent type does make things easier imo.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 3]
```diff
@@ -116,18 +120,56 @@ static bool MultiRouterPlannableQuery(Query *query,
...
+				 RelationRestrictionContext *restrictionContext)
+{
+	if (MultiRouterPlannableQuery(query, restrictionContext))
+	{
+		return CreateSingleTaskRouterPlan(newPlan, originalQuery, query,
+										  restrictionContext);
+	}
+
+	/* FIXME: error? description */
+	return NULL;
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: e.g. why it makes more sense to return a PlanningResult, which could either be NULL itself or have a plannable flag to reflect this situation.

File: src/backend/distributed/planner/multi_planner.c [hunk 4]
```diff
@@ -117,38 +187,154 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
+ */
+static PlannedStmt *
+CreateDistributedPlan(PlannedStmt *localPlan, Query *originalQuery, Query *query,
+					  ParamListInfo boundParams,
+					  RelationRestrictionContext *restrictionContext)
+{
+	MultiPlan *physicalPlan = NULL;
+	PlannedStmt *resultPlan = NULL;
+	DeferredErrorMessage *error = NULL;
+	bool hasUnresolved = false;
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_planner.c]: Alternatively, I've been thinking whether we should call this something like "suppressPlan" (or introduce a second variable with the same value) to make it clearer what happens when this is true.

File: src/backend/distributed/planner/multi_planner.c [hunk 5]
```diff
@@ -117,38 +187,154 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
+		 */
+		error = CreateModifyPlan(&physicalPlan, originalQuery, query,
+								 restrictionContext);
+
+		/*
+		 * Error out if there's an error, unless the error was possibly
+		 * triggered by missing parameters.  In that case an alternative plan
+		 * will be generated with bound parameters (see below).
+		 */
+		if (error && !hasUnresolved)
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_planner.c]: It would be easier to understand what we're doing here if this read: `if (planningError && !suppressPlan)`

File: src/backend/distributed/planner/multi_planner.c [hunk 6]
```diff
@@ -117,38 +187,154 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
-		 * using the extended outfuncs infrastructure, but it's infeasible to
-		 * test most plans. MultiQueryContainerNode always serializes the
-		 * physical plan, so there's no need to check that separately.
+		 * Error out if none of the planners resulted in a usable plan error,
+		 * unless the error was possibly triggered by missing parameters.  In
+		 * that case an alternative plan will be generated with bound
+		 * parameters (see below).
 		 */
-		CheckNodeIsDumpable((Node *) logicalPlan);
+		if (!physicalPlan && error && !hasUnresolved)
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_planner.c]: I think the `!physicalPlan` is redundant here. Also, same comment as for the modify part.

File: src/backend/distributed/executor/multi_executor.c [hunk 7]
```diff
@@ -52,6 +52,19 @@ multi_ExecutorStart(QueryDesc *queryDesc, int eflags)
...
 		Job *workerJob = multiPlan->workerJob;
 
+		/*
+		 * It's possible - particularly in combination with prepared
+		 * statements - that no usable plan has been found. In that case error
+		 * contains the reason for that - raise it.  The reason for
+		 * conditionally erroring out is a slightly hacky approach to
+		 * supporting prepared statements. See
+		 * multi_planner.c:CreateDistributedPlan() for details.
+		 */
```
Comment (marcocitus) [inline src/backend/distributed/executor/multi_executor.c]: I think it's worth elaborating / being more to the point here, e.g. A plan having an error usually means that we did not succeed at planning with unresolved parameters and therefore created a bogus plan with a high cost to make the planner prefer a custom plan. Unfortunately, if the current command is executed from a SQL function then custom plan logic is ignored and we might still try to execute the bogus plan. In this case, we throw the error that was generated during planning.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 8]
```diff
@@ -116,18 +120,56 @@ static bool MultiRouterPlannableQuery(Query *query,
...
+	else
+	{
+		/* modify commands only go through router, even if otherwise disabled */
+		return CreateSingleTaskRouterPlan(newPlan, originalQuery, query,
+										  restrictionContext);
+	}
+}
+
+
+#if 0
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: Obviously this still needs to be moved over to the new functions.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 9]
```diff
@@ -682,18 +747,23 @@ ErrorIfInsertSelectQueryNotSupported(Query *queryTree, RangeTblEntry *insertRte,
...
 		if (!allReferenceTables)
 		{
-			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-							errmsg("If data inserted into a reference table, "
-								   "all of the participating tables in the "
-								   "INSERT INTO ... SELECT query should be "
-								   "reference tables.")));
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "may only reference other reference tables when "
+								 "targeting reference tables with INSERT ... SELECT",
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: Well, the former violates the message guidelines (full sentence, upper case start) and doesn't actually contain more information?

File: src/backend/distributed/planner/multi_planner.c (hunk 4 above)
Comment (anarazel) [inline src/backend/distributed/planner/multi_planner.c]: hasUnresolvedParams sounds good, but suppressPlan would be wrong, because we actually do create some router plans...

File: src/backend/distributed/planner/multi_router_planner.c (hunk 3 above)
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: Well, that'd not really change the issue that it'd be nice to have an error message to report.  We'll right now only show the !router error message. But there's been clear desire to have a guc that only allows router queries, and it'd also be nice if we'd display both error messages when both router and distributed planner fail.  Since we don't really know what query was intended.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 9 above)
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: Maybe reword it a bit to avoid using reference as a verb.

File: src/backend/distributed/planner/multi_planner.c (hunk 4 above)
Comment (anarazel) [inline src/backend/distributed/planner/multi_planner.c]: > Should we still set the cost in that case?

Hm. It'd better if not. I'll move things around so we have a `bool forceCustomPlan = false;`, on which setting the cost to high will depend.

**Comment by anarazel** [comment]
@marcocitus I tried to address most of your comments (some didn't really apply anymore), and do some additional cleanup.  I chose not to add a separate PlanningResult type, but just use the MultiPlan->planningError field added anyway.  Doesn't look too bad imo.

**Comment by marcocitus** [comment]
The following segfaults because workerJob is NULL in MultiExplainOnePlan (trying to explain a bogus plan):

```
CREATE TABLE test (x int, y int);
SELECT create_distributed_table('test', 'x');
PREPARE prep AS SELECT * FROM test WHERE x = $1;
EXPLAIN EXECUTE prep(1);
```

File: src/backend/distributed/planner/multi_router_planner.c [hunk 10]
```diff
@@ -114,67 +116,73 @@ static bool MultiRouterPlannableQuery(Query *query,
...
-		return NULL;
+		return CreateSingleTaskRouterPlan(originalQuery, query,
+										  restrictionContext);
 	}
 
+	/*
+	 * TODO: Instead have MultiRouterPlannableQuery set an error describing
+	 * why router cannot support the query.
+	 */
+	return NULL;
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: For completeness, we could create a MultiPlan with some generic error here.

File: src/backend/distributed/planner/multi_planner.c [hunk 11]
```diff
@@ -122,38 +124,152 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
 	}
 
-	return physicalPlan;
+	/*
+	 * Error out if none of the planners resulted in a usable plan, unless the
+	 * error was possibly triggered by missing parameters.  In that case we'll
+	 * not error out here, but instead rely on postgres' custom plan logic.
+	 * Postgres re-plans prepared statements the first five executions
+	 * (i.e. it produces custom plans), after that the cost of a generic plan
+	 * is compared with the average custom plan cost.  We support otherwise
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_planner.c]: Might be worth pointing out that our fauxFunctionScan always has cost 0.

Comment (marcocitus) [inline src/backend/distributed/planner/multi_planner.c]: or in more detail, that the only cost associated with a custom Citus plan is PostgreSQL's planning cost estimate.

**Comment by marcocitus** [comment]
Other than the explain crash, I wasn't able to find any issues with the approach although I did not try to run a mind-boggling number of queries. I hope it survives 1 or 2 PostgreSQL versions.

File: src/backend/distributed/planner/multi_planner.c [hunk 12]
```diff
@@ -122,38 +124,152 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
+			MultiLogicalPlanOptimize(logicalPlan);
+
+			/*
+			 * This check is here to make it likely that all node types used in
+			 * Citus are dumpable. Explain can dump logical and physical plans
+			 * using the extended outfuncs infrastructure, but it's infeasible to
+			 * test most plans. MultiQueryContainerNode always serializes the
+			 * physical plan, so there's no need to check that separately.
+			 */
+			CheckNodeIsDumpable((Node *) logicalPlan);
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_planner.c]: Yes, I think so - I'd at some point decided to remove the unused output functions for the logical stage, but several people including @metdos and @ozgune said they'd like to keep them.  Having some way to exercise nearly all of them sounds like a good idea.

File: src/backend/distributed/planner/multi_planner.c [hunk 13]
```diff
@@ -122,38 +124,152 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
+						  "could not create distributed plan",
+						  "Possibly this is caused by the use of parameters in SQL "
+						  "functions, which is not supported in Citus.",
+						  "Consider using PLPGSQL functions instead.");
 	}
 
-	return physicalPlan;
+	/*
+	 * Error out if none of the planners resulted in a usable plan, unless the
+	 * error was possibly triggered by missing parameters.  In that case we'll
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_planner.c]: I'm not quite sure what you mean by that?

File: src/backend/distributed/planner/multi_router_planner.c (hunk 10 above)
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: I think short to medium term we should really have router have proper errors for all of  of MultiRouterPlannableQuery (etc's) failures, so we can have a debugging citus.only_router_allowed flag - that'd be quite useful when developing multi tenant applications...

Having some generic error here would probably simplify things, so I'll do that.

**Comment by anarazel** [comment]
> Other than the explain crash, I wasn't able to find any issues with the approach although I did not try to run a mind-boggling number of queries. 

Cool!

>  I hope it survives 1 or 2 PostgreSQL versions.

I don't think there's any current projects that sound like they're going to cause problems, and there've been no problematic changes over the last few releases.  So I'm hopeful.

**Comment by anarazel** [comment]
> The following segfaults because workerJob is NULL in MultiExplainOnePlan (trying to explain a bogus plan):

> CREATE TABLE test (x int, y int);
> SELECT create_distributed_table('test', 'x');
> PREPARE prep AS SELECT * FROM test WHERE x = $1;
> EXPLAIN EXECUTE prep(1);

Besides the issue of not validating whether a plan is valid (fixed + testcase) this afaics fails because explain doesn't deal with parameters correctly (i.e #982).

File: src/backend/distributed/planner/multi_planner.c (hunk 11 above)
Comment (anarazel) [inline src/backend/distributed/planner/multi_planner.c]: I'd rather not expand upon that, I'm about to change that ;)

File: src/backend/distributed/planner/multi_router_planner.c (hunk 10 above)
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: Hm. Looked into it, and there's still several sources of NULLs - RouterSelectTask returning NULL, MultiRouterPlannableQuery returning false, ...

I think I'd prefer opening a ticket to clean this stuff up properly, rather than papering over it with a generic message?

Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: > I think I'd prefer opening a ticket to clean this stuff up properly, rather than papering over it with a generic message?

I.e. have MultiRouterPlannableQuery() return an error (or rather inline into CreateSingleTaskRouterPlan()?), as well as RouterSelectTask/RouterSelectTask() (and a number of subroutines).  That's a good bit of restructuring, so I'd rather not tie this together with prepared statement support.

Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: Good point, let's leave it as is then.

File: src/test/regress/expected/multi_explain.out [hunk 14]
```diff
@@ -727,3 +727,10 @@ Distributed Query into pg_merge_job_570039
 Master Query
   ->  Aggregate
         ->  Seq Scan on pg_merge_job_570039
+-- EXPLAIN EXECUTE of parametrized prepared statements is broken, but
+-- at least make sure to fail without crashing
+PREPARE router_executor_query_param(int) AS SELECT l_quantity FROM lineitem WHERE l_orderkey = $1;
+EXPLAIN EXECUTE router_executor_query_param(5);
+ERROR:  could not create distributed plan
+DETAIL:  Possibly this is caused by the use of parameters in SQL functions, which is not supported in Citus.
```
Comment (marcocitus) [inline src/test/regress/expected/multi_explain.out]: The detail is rather confusing. Could we print a generic error in EXPLAIN if the plan is invalid?

**Comment by anarazel** [comment]
On January 22, 2017 11:20:07 PM PST, Marco Slot <notifications@github.com> wrote:
>marcocitus commented on this pull request.
>
>
>
>> @@ -727,3 +727,10 @@ Distributed Query into pg_merge_job_570039
> Master Query
>   ->  Aggregate
>         ->  Seq Scan on pg_merge_job_570039
>+-- EXPLAIN EXECUTE of parametrized prepared statements is broken, but
>+-- at least make sure to fail without crashing
>+PREPARE router_executor_query_param(int) AS SELECT l_quantity FROM
>lineitem WHERE l_orderkey = $1;
>+EXPLAIN EXECUTE router_executor_query_param(5);
>+ERROR:  could not create distributed plan
>+DETAIL:  Possibly this is caused by the use of parameters in SQL
>functions, which is not supported in Citus.
>
>The detail is rather confusing. Could we print a generic error in
>EXPLAIN if the plan is invalid?

We could, but I don't think it'll matter much. The likelihood of somebody hitting this seems low, and I'm going to open apr with explain support for prepared statements soon (needs some cleanup, otherwise done).  You want me to?
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
